File: devkit-lib/shell.py
#!/usr/bin/env python3
"""devkit interactive shell — fully self-contained, zero deps."""

import os
import subprocess
import sys
import termios
import tty
from typing import List, Tuple


DEVKIT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))

TOOLS: List[Tuple[str, str]] = [
    ("totp",    "TOTP 2FA code manager"),
    ("curl",    "HTTP client (Postman-like)"),
    ("docker",  "Docker container manager"),
    ("cleanup", "Disk space cleanup"),
    ("vpn",     "VPN connection manager"),
]

CYAN = '\033[36m'
BOLD = '\033[1m'
NC = '\033[0m'


def _set_vmin_vtime(fd: int, vmin: int, vtime: int):
    """Set termios VMIN and VTIME for a file descriptor."""
    attrs = termios.tcgetattr(fd)
    attrs[6][termios.VMIN] = vmin
    attrs[6][termios.VTIME] = vtime
    termios.tcsetattr(fd, termios.TCSANOW, attrs)


def read_key(fd: int) -> str:
    """Read a single keypress in raw mode. Returns UP/DOWN/ENTER/ESC or the character."""

    char = sys.stdin.read(1)

    if char == '\x1b':
        # Escape sequence: set 100ms timeout (VTIME=1 decisecond, VMIN=0)
        _set_vmin_vtime(fd, 0, 1)
        seq = sys.stdin.read(2) if sys.stdin.readable() else ''
        # Restore blocking mode
        _set_vmin_vtime(fd, 1, 0)

        if seq in ('[A', 'OA'):
            return 'UP'
        elif seq in ('[B', 'OB'):
            return 'DOWN'
        elif not seq or seq is None:
            return 'ESC'
        return ''

    if char in ('\r', '\n'):
        return 'ENTER'
    if char == '\x7f':
        return 'BACKSPACE'
    return char


def show_suggestions(filter_text: str) -> Tuple[str, int]:
    """Show interactive suggestion popup. Returns (tool_name, index) or ('', -1)."""
    items = [(n, d) for n, d in TOOLS if filter_text in n]
    if not items:
        return '', -1

    selected = 0
    count = len(items)
    lines = count + 2

    fd = sys.stdin.fileno()
    old = termios.tcgetattr(fd)
    tty.setraw(fd)

    try:
        sys.stderr.write('\033[?25l')
        sys.stderr.flush()

        while True:
            for i in range(count):
                prefix = ' >' if i == selected else '  '
                line = f'\r  {prefix}  {CYAN}{items[i][0]:<10}{NC}  {items[i][1]:<28}\033[J\n'
                sys.stderr.write(line)

            sys.stderr.write(f'\r  {BOLD}↑↓{NC} nav  {BOLD}Enter{NC} select  {BOLD}Esc{NC} cancel\033[J\n')
            sys.stderr.write(f'\033[{lines}A')
            sys.stderr.flush()

            key = read_key(fd)

            if key == 'UP':
                selected = (selected - 1) % count
            elif key == 'DOWN':
                selected = (selected + 1) % count
            elif key == 'ENTER':
                return items[selected][0], selected
            elif key == 'ESC':
                return '', -1
            elif len(key) == 1 and key.isprintable():
                filter_text += key
                items = [(n, d) for n, d in TOOLS if filter_text in n]
                if not items:
                    return key, -2
                selected = 0
                count = len(items)
                lines = count + 2
    finally:
        sys.stderr.write(f'\033[{lines}B\033[J\033[?25h')
        sys.stderr.flush()
        termios.tcsetattr(fd, termios.TCSADRAIN, old)

    return '', -1


def run_interactive():
    """Main interactive REPL."""
    os.chdir(DEVKIT_DIR)

    print()
    print(f'  {BOLD}╭────────────────────────────────╮{NC}')
    print(f'  {BOLD}│{NC}          {BOLD}devkit{NC} v1.0            {BOLD}│{NC}')
    print(f'  {BOLD}│{NC}     Smart CLI Toolbox           {BOLD}│{NC}')
    print(f'  {BOLD}╰────────────────────────────────╯{NC}')
    print()

    while True:
        try:
            cmd = input('  devkit> ').strip()
        except (EOFError, KeyboardInterrupt):
            print()
            print(f'  {CYAN}Goodbye!{NC}')
            break

        if not cmd:
            print()
            print(f'  {BOLD}Available tools:{NC}')
            for name, desc in TOOLS:
                print(f'  {CYAN}/{name:<10}{NC}  {desc}')
            print(f'  Type {BOLD}/<tool>{NC} or a name. {BOLD}q{NC} to quit.')
            print()
            continue

        if cmd in ('q', 'quit', 'exit', '0'):
            print()
            print(f'  {CYAN}Goodbye!{NC}')
            break

        if cmd.startswith('/'):
            filter_text = cmd[1:].strip()

            # Quick match
            for name, _ in TOOLS:
                if name.startswith(filter_text) and len(filter_text) > 0:
                    run_tool(name)
                    print()
                    break
            else:
                tool_name, _ = show_suggestions(filter_text)
                if tool_name:
                    run_tool(tool_name)
                    print()
            continue

        dispatch = {
            '1': 'totp', '2': 'curl', '3': 'docker', '4': 'cleanup', '5': 'vpn',
            'totp': 'totp', 'curl': 'curl', 'docker': 'docker', 'cleanup': 'cleanup', 'vpn': 'vpn',
        }
        if cmd in dispatch:
            run_tool(dispatch[cmd])
            print()
        else:
            print()
            print(f'  \033[31m[ERR]\033[0m  Unknown: {cmd}')
            print()


def run_tool(name: str):
    """Launch a devkit tool in a subprocess."""
    devkit = os.path.join(DEVKIT_DIR, 'devkit')
    print()
    subprocess.run([devkit, name])


def main():
    run_interactive()


if __name__ == '__main__':
    main()

File: devkit-lib/shell.js
#!/usr/bin/env node
/**
 * devkit interactive shell — Node.js
 * Real-time suggestion popup on "/" like Claude Code.
 * Uses raw stdin for full control.
 */
const { execFileSync } = require('child_process');
const path = require('path');

const DEVKIT_DIR = path.resolve(__dirname, '..');
const DEVKIT_BIN = path.join(DEVKIT_DIR, 'devkit');

const TOOLS = [
  { name: 'totp',    desc: 'TOTP 2FA code manager' },
  { name: 'curl',    desc: 'HTTP client (Postman-like)' },
  { name: 'docker',  desc: 'Docker container manager' },
  { name: 'cleanup', desc: 'Disk space cleanup' },
  { name: 'vpn',     desc: 'VPN connection manager' },
];

const CYAN = '\x1b[36m', BOLD = '\x1b[1m', NC = '\x1b[0m';
const HIDE = '\x1b[?25l', SHOW = '\x1b[?25h', CLR = '\x1b[J';


/** Read N bytes from stdin with timeout (ms). Throws on timeout. */
function readBytes(n, timeoutMs = 100) {
  return new Promise((resolve, reject) => {
    const timer = setTimeout(() => reject(new Error('timeout')), timeoutMs);
    function attempt() {
      const data = process.stdin.read(n);
      if (data && data.length >= n) {
        clearTimeout(timer);
        resolve(data.slice(0, n));
        if (data.length > n) process.stdin.unshift(data.slice(n));
      } else {
        process.stdin.once('readable', attempt);
      }
    }
    process.stdin.once('readable', attempt);
  });
}

/** Read one keypress. Returns 'a', '\r', '\x1b[A', '\x1bOA', '\x1b' etc. */
async function readKey(timeoutMs = 5000) {
  const first = await readBytes(1, timeoutMs);
  const b = first[0];

  if (b === 0x1b) {
    try {
      const second = await readBytes(1, 30);
      if (second[0] === 0x5b || second[0] === 0x4f) {
        const third = await readBytes(1, 30);
        return `\x1b${String.fromCharCode(second[0])}${String.fromCharCode(third[0])}`;
      }
      return '\x1b' + String.fromCharCode(second[0]);
    } catch {
      return '\x1b'; // Just Escape key
    }
  }

  if (b === 0x7f) return '\x7f';
  return String.fromCharCode(b);
}


/** Show interactive suggestion popup. Returns tool name or null. */
async function suggestPopup(filterText) {
  let items = TOOLS.filter(t => t.name.includes(filterText));
  if (items.length === 0) return null;

  let selected = 0;
  let count = items.length;
  let lines = count + 2;

  process.stderr.write(HIDE);

  while (true) {
    for (let i = 0; i < count; i++) {
      const p = i === selected ? ' >' : '  ';
      process.stderr.write(`\r  ${p}  ${CYAN}${items[i].name.padEnd(10)}${NC}  ${items[i].desc.padEnd(28)}${CLR}\n`);
    }
    process.stderr.write(`\r  ${BOLD}↑↓${NC} nav  ${BOLD}Enter${NC} select  ${BOLD}Esc${NC} cancel${CLR}\n`);
    process.stderr.write(`\x1b[${lines}A`);
    process.stderr.flush?.();

    const key = await readKey(5000);

    if (key === '\x1b[A' || key === '\x1bOA') {
      selected = (selected - 1 + count) % count;
    } else if (key === '\x1b[B' || key === '\x1bOB') {
      selected = (selected + 1) % count;
    } else if (key === '\r' || key === '\n') {
      process.stderr.write(`\x1b[${lines}B${CLR}${SHOW}`);
      return items[selected].name;
    } else if (key === '\x1b') {
      process.stderr.write(`\x1b[${lines}B${CLR}${SHOW}`);
      return null;
    } else if (key.length === 1 && key >= ' ') {
      filterText += key;
      items = TOOLS.filter(t => t.name.includes(filterText));
      if (items.length === 0) {
        process.stderr.write(`\x1b[${lines}B${CLR}${SHOW}`);
        return key; // pass char back
      }
      count = items.length;
      lines = count + 2;
      selected = 0;
    }
  }
}


function runTool(name) {
  console.log();
  try { execFileSync(DEVKIT_BIN, [name], { stdio: 'inherit' }); } catch {}
  console.log();
}


async function main() {
  if (!process.stdin.isTTY) {
    process.stderr.write('devkit interactive shell requires a TTY\n');
    process.exit(1);
  }

  process.stdin.setRawMode(true);
  process.stdin.resume();

  console.log();
  console.log(`  ${BOLD}╭────────────────────────────────╮${NC}`);
  console.log(`  ${BOLD}│${NC}          ${BOLD}devkit${NC} v1.0            ${BOLD}│${NC}`);
  console.log(`  ${BOLD}│${NC}     Smart CLI Toolbox           ${BOLD}│${NC}`);
  console.log(`  ${BOLD}╰────────────────────────────────╯${NC}`);
  console.log();

  let line = '';
  process.stdout.write('  devkit> ');

  while (true) {
    const key = await readKey(5000);

    // Ctrl+C
    if (key === '\x03') {
      console.log(`\n  ${CYAN}Goodbye!${NC}`);
      break;
    }

    // Enter
    if (key === '\r' || key === '\n') {
      const cmd = line.trim();
      line = '';
      process.stdout.write('\n');

      if (!cmd) {
        console.log(`\n  ${BOLD}Available tools:${NC}`);
        for (const t of TOOLS) console.log(`  ${CYAN}/${t.name.padEnd(10)}${NC}  ${t.desc}`);
        console.log(`  Type ${BOLD}/<tool>${NC} or a name. ${BOLD}q${NC} to quit.\n`);
        process.stdout.write('  devkit> ');
        continue;
      }

      if (['q', 'quit', 'exit', '0'].includes(cmd)) {
        console.log(`  ${CYAN}Goodbye!${NC}`);
        break;
      }

      if (cmd === '/') {
        const result = await suggestPopup('');
        if (result && typeof result === 'string') runTool(result);
      } else if (cmd.startsWith('/')) {
        const filter = cmd.slice(1).trim();
        const match = TOOLS.find(t => t.name.startsWith(filter) && filter.length > 0);
        match ? runTool(match.name) : (await suggestPopup(filter));
      } else {
        const d = {
          '1': 'totp', '2': 'curl', '3': 'docker', '4': 'cleanup', '5': 'vpn',
          totp: 'totp', curl: 'curl', docker: 'docker', cleanup: 'cleanup', vpn: 'vpn',
        };
        if (d[cmd]) { runTool(d[cmd]); }
        else { console.log(`  \x1b[31m[ERR]${NC}  Unknown: ${cmd}\n`); }
      }

      process.stdout.write('  devkit> ');
      continue;
    }

    // Backspace
    if (key === '\x7f' || key === '\b') {
      if (line.length > 0) { line = line.slice(0, -1); process.stdout.write('\b \b'); }
      continue;
    }

    // Printable char
    if (key.length === 1 && key >= ' ') {
      line += key;
      process.stdout.write(key);

      // Immediate popup on standalone "/"
      if (line === '/') {
        process.stdout.write('\n');
        const result = await suggestPopup('');
        if (result && typeof result === 'string') runTool(result);
        line = '';
        process.stdout.write('  devkit> ');
      }
    }
  }

  process.stdin.setRawMode(false);
}

main().catch(() => process.exit(1));
